pythran/pythonic/types/variant_functor.hpp

```diff
@@ -0,0 +1,139 @@
+#ifndef PYTHONIC_TYPES_VARIANT_FUNCTOR_HPP
+#define PYTHONIC_TYPES_VARIANT_HPP
+
+#include "pythonic/include/types/variant_functor.hpp"
+
+#include <utility>
+#include <iostream>
+
+namespace pythonic
+{
+
+  namespace types
+  {
+
+    namespace details
+    {
+
+      template <class Type>
+      variant_functor_impl<Type>::variant_functor_impl(char mem[],
+                                                       Type const &t)
+          : head(new (mem) Type(t))
+      {
+      }
+
+      template <class Type>
+      template <class OtherType>
+      variant_functor_impl<Type>::variant_functor_impl(char mem[],
+                                                       OtherType const &t)
+          : head(nullptr)
+      {
+      }
+
+      template <class Type>
+      variant_functor_impl<Type>::variant_functor_impl(
+          char mem[], variant_functor_impl<Type> const &t)
+          : head(t.head ? new (mem) Type(*t.head) : nullptr)
+      {
+      }
+
+      template <class Type>
+      template <class... OtherTypes>
+      variant_functor_impl<Type>::variant_functor_impl(
+          char mem[], variant_functor_impl<Type, OtherTypes...> const &t)
+          : variant_functor_impl(mem, t.head)
+      {
+      }
+
+      template <class Type>
+      template <class OtherType, class... OtherTypes>
+      variant_functor_impl<Type>::variant_functor_impl(
+          char mem[], variant_functor_impl<OtherType, OtherTypes...> const &t)
+          : variant_functor_impl(mem, t.tail)
+      {
+      }
+
+      template <class Type>
+      template <class... Args>
+      auto variant_functor_impl<Type>::operator()(Args &&... args)
+          -> decltype(std::declval<Type>()(std::forward<Args>(args)...))
+      {
+        return (*head)(std::forward<Args>(args)...);
+      }
+
+      template <class Type>
+      template <class... Args>
+      auto variant_functor_impl<Type>::operator()(Args &&... args) const
+          -> decltype(std::declval<Type>()(std::forward<Args>(args)...))
+      {
+        return (*head)(std::forward<Args>(args)...);
+      }
+
+      template <class Type, class... Types>
+      template <class... OtherTypes>
+      variant_functor_impl<Type, Types...>::variant_functor_impl(
+          char mem[], OtherTypes const &... t)
+          : head(mem, t...), tail(mem, t...)
```

Comment:
I think in that case there will be two head that points to the functor, but that should be ok